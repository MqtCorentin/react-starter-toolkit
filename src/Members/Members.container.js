import React from 'react';
import PropTypes from 'prop-types';
import { LoaderModes } from '@axa-fr/react-toolkit-all';
import withEnhancedCustomFetch from 'shared/hoc/withCustomFetch';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import { useNotifications } from 'shared/components/Notifications';
import Members from './Members';
import { useMembers } from './Members.hook';

export const MembersContext = React.createContext({ onChangeSorting: null, sorting: {}, addNotification: null, stateNotifications: [] });
const { Provider: MembersProvider } = MembersContext;

export const MembersEnhanced = ({ useMembersFn, useNotificationsFn, setLoaderModeFn, fetchCustom, ...rest }) => {
  const { anomaly, isLoading, members, pagination, onChangeSorting, stateSorting, onChangePaging } = useMembersFn({ fetchCustom });
  const { addNotification, onDeleteNotification, stateNotifications } = useNotificationsFn();

  return (
    <MembersProvider value={{ onChangeSorting, sorting: stateSorting, addNotification, stateNotifications }}>
      <Members
        {...rest}
        members={members}
        loaderMode={setLoaderModeFn({ isLoading, LoaderModes })}
        anomaly={anomaly}
        deleteNotification={onDeleteNotification}
        notifications={stateNotifications}
        pagination={pagination}
        onChangePaging={onChangePaging}
      />
    </MembersProvider>
  );
};

MembersEnhanced.propTypes = {
  useMembersFn: PropTypes.func,
  useNotificationsFn: PropTypes.func,
  setLoaderModeFn: PropTypes.func,
};

MembersEnhanced.defaultProps = {
  useMembersFn: useMembers,
  useNotificationsFn: useNotifications,
  setLoaderModeFn: setLoaderMode,
};

export default withEnhancedCustomFetch(MembersEnhanced);
