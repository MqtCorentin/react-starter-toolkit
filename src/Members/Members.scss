@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';
@import '../shared/scss/mixins.scss';

.af {
  &-btn {
    &__img {
      width: 19px;
      fill: $white;
      @include centerer;
    }
    &--circle-small {
      &:hover {
        .af {
          &-btn {
            &__img {
              fill: $color-axa;
            }
          }
        }
      }
    }
    &--circle-disabled {
      background-color: $color-gray-2;
      border-color: $color-gray-1;
      cursor: not-allowed;
      color: $color-gray-1;
      &:hover {
        background-color: $color-gray-2;
        border-color: $color-gray-1;
        color: $color-gray-1;
      }
    }
  }

  &-actions {
    display: flex;
    justify-content: space-evenly;
  }
}
