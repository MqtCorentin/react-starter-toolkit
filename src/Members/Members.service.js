export default ({ fetchCustom, signal, field = 'lastname', order = -1, max = 50, skip = 0 }) =>
  fetchCustom(`members?max=${max}&sort=${field}&dir=${order}&skip=${skip}`, { signal });
