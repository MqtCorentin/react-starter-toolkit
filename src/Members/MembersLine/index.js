export { default } from './MembersLine.container';
