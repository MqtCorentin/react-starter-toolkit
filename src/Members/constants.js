export const TITLE_BAR = 'Sociétaires';
export const TITLE = 'Liste des sociétaires';

export const FETCH_MEMBERS_INIT = 'FETCH_MEMBERS_INIT';
export const FETCH_MEMBERS_SUCCESS = 'FETCH_MEMBERS_SUCCESS';
export const FETCH_MEMBERS_FAILURE = 'FETCH_MEMBERS_FAILURE';
export const FETCH_MEMBERS_ORDER = 'FETCH_MEMBERS_ORDER';

export const TABLE_HEADERS_MEMBERS = [
  { label: 'Prénom', field: 'firstname' },
  { label: 'Nom', field: 'lastname' },
  { label: 'Date de naissance', field: 'birthdate' },
  { label: 'Sexe', field: 'sexe' },
];
