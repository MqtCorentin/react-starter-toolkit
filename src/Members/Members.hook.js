import { useEffect, useReducer, useCallback } from 'react';
import moment from 'moment';
import { get } from 'lodash';
import { NONE } from 'shared/constants';
import fetchData from 'shared/helpers/fetchData';
import findMembers from './Members.service';
import { FETCH_MEMBERS } from './constants';

const initState = {
  isLoading: false,
  members: [],
  anomaly: null,
  sorting: {
    field: '',
    order: NONE,
  },
  filters: {
    numberItems: 10,
    currentPage: 1,
    numberPages: 1,
  },
};

export const computeInfos = ({ members = [], getFn = get, momentFn = moment }) =>
  members.map(member => ({
    ...member,
    birthdate: momentFn(getFn(member, 'birthdate')).format('DD/MM/YYYY'),
  }));

export const setStateMembersLoading = ({ state }) => ({
  ...state,
  isLoading: true,
});

export const setStateMembersSuccess = ({ state, payload, computeInfosFn = computeInfos }) => ({
  ...state,
  isLoading: false,
  members: computeInfosFn({
    members: payload.members.responseBody,
  }),
});

export const setStateMembersFailure = ({ state, payload }) => ({
  ...state,
  isLoading: false,
  anomaly: payload.members,
});

export const setStateMembersOrder = ({ state, payload }) => ({
  ...state,
  sorting: {
    ...state.sorting,
    ...payload,
  },
  isLoading: true,
});

export const dataFetchReducer = (
  state,
  {
    type,
    payload,
    setStateMembersLoadingFn = setStateMembersLoading,
    setStateMembersSuccessFn = setStateMembersSuccess,
    setStateMembersFailureFn = setStateMembersFailure,
    setStateMembersOrderFn = setStateMembersOrder,
  },
) => {
  switch (type) {
    case FETCH_MEMBERS.INIT:
      return setStateMembersLoadingFn({ state });
    case FETCH_MEMBERS.SUCCESS:
      return setStateMembersSuccessFn({ state, payload });
    case FETCH_MEMBERS.ORDER:
      return setStateMembersOrderFn({ state, payload });
    case FETCH_MEMBERS.FAILURE:
      return setStateMembersFailureFn({ state, payload });
    default:
      return state;
  }
};

export const setMembersInit = dispatch => () => dispatch({ type: FETCH_MEMBERS.INIT });
export const setMembersError = dispatch => payload => dispatch({ type: FETCH_MEMBERS.FAILURE, payload });
export const setMembersSuccess = dispatch => payload => dispatch({ type: FETCH_MEMBERS.SUCCESS, payload });
export const setMembersOrder = dispatch => payload => dispatch({ type: FETCH_MEMBERS.ORDER, payload });

export const useMembers = ({
  fetchCustom,
  initStateCt = initState,
  fetchDataFn = fetchData,
  dataFetchReducerFn = dataFetchReducer,
  findMembersFn = findMembers,
  setMembersInitFn = setMembersInit,
  setMembersErrorFn = setMembersError,
  setMembersSuccessFn = setMembersSuccess,
  setMembersOrderFn = setMembersOrder,
}) => {
  const [stateMembers, dispatch] = useReducer(dataFetchReducerFn, initStateCt);
  const {
    sorting: { field, order },
  } = stateMembers;

  const onChangeOrder = useCallback(sorting => setMembersOrderFn(dispatch)(sorting), [setMembersOrderFn]);

  useEffect(() => {
    fetchDataFn({
      fetchCustom,
      setInit: setMembersInitFn(dispatch),
      setError: setMembersErrorFn(dispatch),
      setSuccess: setMembersSuccessFn(dispatch),
      fetchServices: {
        members: {
          service: findMembersFn,
          args: {
            field,
            order,
          },
        },
      },
    });
  }, [fetchCustom, fetchDataFn, field, findMembersFn, order, setMembersErrorFn, setMembersInitFn, setMembersSuccessFn]);

  return { ...stateMembers, onChangeOrder };
};
