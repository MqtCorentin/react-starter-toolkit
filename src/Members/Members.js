import React from 'react';
import Resilience from 'shared/components/Resilience/Resilience';
import { Paging } from '@axa-fr/react-toolkit-all';
import Loader from 'shared/components/Loader';
import Table from 'shared/components/Table';
import { TITLE_BAR, TITLE, TABLE_HEADERS_MEMBERS } from './constants';

const Members = ({ header, footer, title, menu, loaderMode, members, anomaly, onChangePaging, pagination, onChangeSorting, sorting }) => (
  <>
    {header()}
    {menu()}
    {title({ title: TITLE_BAR, backHome: true })}
    <div className="af-main container">
      <h1 className="af-title--content">{TITLE}</h1>
      <Loader mode={loaderMode}>
        <Resilience anomaly={anomaly}>
          <Table items={members} headers={TABLE_HEADERS_MEMBERS} onSort={onChangeSorting} sorting={sorting} />
          <Paging {...pagination} onChange={onChangePaging} id="paging" />
        </Resilience>
      </Loader>
    </div>
    {footer()}
  </>
);

export default Members;
