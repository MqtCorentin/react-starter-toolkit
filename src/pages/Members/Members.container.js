import React, { createContext } from 'react';
import { func } from 'prop-types';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import Members from './Members';
import { useMembers } from './Members.hook';
import { SERVICE_NAME } from './constants';

export const MembersContext = createContext({ onChangeSorting: null, sorting: {} });
const { Provider: MembersProvider } = MembersContext;

const MembersEnhanced = ({ useMembersFn, setLoaderModeFn, ...rest }) => {
  const { anomaly, isLoading, members, onChangeSorting, stateSorting, onChangePaging } = useMembersFn({});
  return (
    <MembersProvider value={{ onChangeSorting, sorting: stateSorting }}>
      <Members
        {...rest}
        members={members.data}
        loaderMode={setLoaderModeFn({ isLoading })}
        anomaly={anomaly[SERVICE_NAME]}
        pagination={members.pagination}
        onChangePaging={onChangePaging}
        onChangeSorting={onChangeSorting}
        sorting={stateSorting}
      />
    </MembersProvider>
  );
};

MembersEnhanced.propTypes = {
  useMembersFn: func,
  setLoaderModeFn: func,
};

MembersEnhanced.defaultProps = {
  useMembersFn: useMembers,
  setLoaderModeFn: setLoaderMode,
};

export default MembersEnhanced;
