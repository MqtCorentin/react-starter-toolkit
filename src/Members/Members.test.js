import React from 'react';
import { render } from '@testing-library/react';
import Members, { MembersItems } from './Members';

const defaultProps = {
  header: () => {},
  title: () => {},
  footer: () => {},
  members: [],
  anomaly: null,
  loaderMode: 'none',
  deleteNotification: () => {},
  notifications: [],
};

describe('<Members/>', () => {
  it('Should render Members', () => {
    const { asFragment } = render(<Members {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<MembersItems/>', () => {
  it('Should render MembersItems', () => {
    const { asFragment } = render(<MembersItems {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should contain 1 <Table.Tr /> element when array members contains 1 member', () => {
    const { asFragment, container, getByRole } = render(
      <MembersItems
        {...defaultProps}
        members={[
          {
            _id: '00001',
            type: 'PP',
            firstname: 'Samuel',
            lastname: 'Gomez',
            birthdate: '1985-10-20T13:44:20.540000',
            created: '2020-10-20T13:44:20.540000',
          },
        ]}
      />,
    );
    expect(getByRole('table')).not.toBeNull();
    expect(container.querySelector('.af-alert.af-alert--notification.af-alert--error')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<Members/> with <Notification />', () => {
  it('Should render Members with <Notification /> when have notifications', () => {
    const { asFragment, container } = render(
      <Members {...defaultProps} notifications={[{ id: '001', code: 'notif', title: 'title', onClose: () => {} }]} />,
    );

    expect(container.querySelector('.af-alert.af-alert--notification.af-alert--error')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
});
