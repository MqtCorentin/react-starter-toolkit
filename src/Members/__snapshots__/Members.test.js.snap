// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Members/> Should render Members 1`] = `
<DocumentFragment>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Liste des membres
    </h1>
    <div
      style="min-height: 94px; position: relative;"
    >
      <div
        class="af-paging"
      >
        <div
          class="af-paging__limit"
        >
          <div
            class="af-paging__limit"
          >
            <form
              class="af-paging__form"
            >
              <div
                class="af-form__group"
              >
                <div
                  class="col col-sm-2 col-md-2 col-lg-2 col-xl-2"
                >
                  <label
                    class="af-form__group-label"
                    for="paging"
                  >
                    Afficher
                  </label>
                </div>
                <div
                  class="col col-sm-10 col-md-10 col-lg-10 col-xl-10"
                >
                  <div
                    class="af-form__select"
                  >
                    <div
                      class="af-form__select-container"
                    >
                      <select
                        class="af-form__input-select"
                        id="paging"
                      >
                        <option
                          value="5"
                        >
                          5
                        </option>
                        <option
                          value="10"
                        >
                          10
                        </option>
                        <option
                          value="25"
                        >
                          25
                        </option>
                        <option
                          value="50"
                        >
                          50
                        </option>
                        <option
                          value="100"
                        >
                          100
                        </option>
                      </select>
                      <span
                        class="glyphicon glyphicon-menu-down"
                      />
                    </div>
                    <span
                      class="af-form__input-cmplt"
                    >
                      éléments
                    </span>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="af-paging__pager"
        >
          <nav
            class="af-pager"
          >
            <ul
              class="af-pager__pagination"
            >
              <li
                class="af-pager__item af-pager__item--disabled"
              >
                <span
                  class="af-pager__item-nolink"
                >
                  « Précédent
                </span>
              </li>
              <li
                class="af-pager__item af-pager__item--active"
              >
                <a
                  class="af-pager__item-link"
                  href="#"
                >
                  <span>
                    1
                  </span>
                </a>
              </li>
              <li
                class="af-pager__item af-pager__item--disabled"
              >
                <span
                  class="af-pager__item-nolink"
                >
                  Suivant »
                </span>
              </li>
            </ul>
          </nav>
        </div>
      </div>
       
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Members/> with <Notification /> Should render Members with <Notification /> when have notifications 1`] = `
<DocumentFragment>
  <aside
    class="af-notifications af-notifications--open"
  >
    <div
      class="af-alert af-alert--notification af-alert--error"
    >
      <button
        class="af-alert__icon-close"
      >
        <i
          class="glyphicon glyphicon-close"
        />
      </button>
      <div
        class="af-alert__title"
      >
        <div
          class="af-alert__title-icon"
        >
          <i
            class="glyphicon glyphicon-minus-sign"
          />
        </div>
        <div
          class="af-alert__title-text"
        >
          title
        </div>
      </div>
    </div>
  </aside>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Liste des membres
    </h1>
    <div
      style="min-height: 94px; position: relative;"
    >
      <div
        class="af-paging"
      >
        <div
          class="af-paging__limit"
        >
          <div
            class="af-paging__limit"
          >
            <form
              class="af-paging__form"
            >
              <div
                class="af-form__group"
              >
                <div
                  class="col col-sm-2 col-md-2 col-lg-2 col-xl-2"
                >
                  <label
                    class="af-form__group-label"
                    for="paging"
                  >
                    Afficher
                  </label>
                </div>
                <div
                  class="col col-sm-10 col-md-10 col-lg-10 col-xl-10"
                >
                  <div
                    class="af-form__select"
                  >
                    <div
                      class="af-form__select-container"
                    >
                      <select
                        class="af-form__input-select"
                        id="paging"
                      >
                        <option
                          value="5"
                        >
                          5
                        </option>
                        <option
                          value="10"
                        >
                          10
                        </option>
                        <option
                          value="25"
                        >
                          25
                        </option>
                        <option
                          value="50"
                        >
                          50
                        </option>
                        <option
                          value="100"
                        >
                          100
                        </option>
                      </select>
                      <span
                        class="glyphicon glyphicon-menu-down"
                      />
                    </div>
                    <span
                      class="af-form__input-cmplt"
                    >
                      éléments
                    </span>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="af-paging__pager"
        >
          <nav
            class="af-pager"
          >
            <ul
              class="af-pager__pagination"
            >
              <li
                class="af-pager__item af-pager__item--disabled"
              >
                <span
                  class="af-pager__item-nolink"
                >
                  « Précédent
                </span>
              </li>
              <li
                class="af-pager__item af-pager__item--active"
              >
                <a
                  class="af-pager__item-link"
                  href="#"
                >
                  <span>
                    1
                  </span>
                </a>
              </li>
              <li
                class="af-pager__item af-pager__item--disabled"
              >
                <span
                  class="af-pager__item-nolink"
                >
                  Suivant »
                </span>
              </li>
            </ul>
          </nav>
        </div>
      </div>
       
    </div>
  </div>
</DocumentFragment>
`;
